python/lsst/daf/butler/core/dbAuth.py

```diff
@@ -0,0 +1,227 @@
+        * "postgresql://*.example.com"
+        * "postgresql://*.example.com/my_*"
+        * "postgresql://host.example.com/my_database"
+        * "postgresql://host.example.com:5432/my_database"
+        * "postgresql://user@host.example.com/my_database"
+
+        Note that the connection URL
+        "postgresql://host.example.com/my_database" would not match against
+        the pattern "postgresql://host.example.com:5432", even if the default
+        port for the connection is 5432.
+        """
+        if drivername is None or drivername == "":
+            raise DbAuthError("Missing drivername parameter")
+        if host is None or host == "":
+            raise DbAuthError("Missing host parameter")
+        if database is None or database == "":
+            raise DbAuthError("Missing database parameter")
+
+        for authDict in self.authList:
+
+            # Check for mandatory entries
+            if "url" not in authDict:
+                raise DbAuthError("Missing URL in DbAuth configuration")
+            if "password" not in authDict:
+                raise DbAuthError(
+                    "Missing password in DbAuth configuration "
+                    "for URL: " + authDict["url"])
+
+            # Parse pseudo-URL from db-auth.yaml
+            components = urllib.parse.urlparse(authDict["url"])
+
+            # Check for same database backend type/driver
+            if components.scheme == "":
+                raise DbAuthError(
+                    "Missing database driver in URL: " + authDict["url"])
+            if drivername != components.scheme:
+                continue
+
+            # Check for same database name
+            if components.path != "" and components.path != "/":
+                if not fnmatch.fnmatch(database, components.path.lstrip("/")):
+                    continue
+
+            # Parse out user if present and check
+            if components.netloc == "":
+                raise DbAuthError("Missing host in URL: " + authDict["url"])
+            if "@" in components.netloc:
+                matchUser, matchHost = components.netloc.split("@")
+                if (username is None or username == "") and matchUser != "":
+                    continue
+                if username is not None and username != matchUser:
+                    continue
+            else:
+                matchHost = components.netloc
+
+            # Parse out host and port and check
+            matchPort = None
+            # Check for IP v6
+            if matchHost.startswith("["):
+                match = re.search(r"^(\[[0-9a-f:]+\]):(\d+)$", matchHost)
+                if match:
+                    matchHost, matchPort = match.groups()
+                # Prep for fnmatch
+                matchHost = re.sub(r"^\[", "[[]", matchHost)
+            elif ":" in matchHost:
+                matchHost, matchPort = matchHost.split(":")
+            if matchPort is not None and (port is None or port != matchPort):
```

Comment:
I'm not sure if this will work if port is an `int` because `matchPort` is always an `str`. I know the documentation states that `port` must be `str` but would it be possible to make a `str()` cast here and relax that requirement?

Reply:
What do you get out of SQLAlchemy?

Reply:
I think sqlalchemy would return me an `str` while `urlib.parse.urlparse("postgresql://test:testpass@host:1234/dbname")` would return an `int` and explicitly defined keys in the butler yaml config would depend whether or not the given port was wrapped in `""` or not

Reply:
Changed to enable numeric ports.